### ISSUE TITLE
Dispose called multiple time for same instance when registering service for multiple interfaces

### ISSUE THREAD

**Issue by pihalve** [aspnet/Extensions#1155]
### Describe the bug
I actually don't know if it's a bug or by design, but to me it looks like a bug. Basically, what I'm trying to do is registering same singleton instance as multiple interfaces. I know that there is not explicit support for this in M.E.DI, but it can be sort of accomplished by doing this:

    services.AddSingleton<Foo>(); // We must explicitly register Foo
    services.AddSingleton<IFoo>(x => x.GetRequiredService<Foo>()); // Forward requests to Foo
    services.AddSingleton<IBar>(x => x.GetRequiredService<Foo>()); // Forward requests to Foo

The same instance of Foo can now be injected as both IFoo and IBar. However, the problem is, that if Foo implements IDisposable, then Dispose will be called for all three resolved registrations, i.e. for Foo, IFoo and IBar.

### To Reproduce
1. Create a class called Foo that implements two interfaces IFoo and IBar. Additionally, have Foo implement IDisposable.

2. Do registrations like this:
    services.AddSingleton<Foo>(); // We must explicitly register Foo
    services.AddSingleton<IFoo>(x => x.GetRequiredService<Foo>()); // Forward requests to Foo
    services.AddSingleton<IBar>(x => x.GetRequiredService<Foo>()); // Forward requests to Foo

2. Resolve the registrations:
    var foo1 = provider.GetService<Foo>(); // An instance of Foo
    var foo2 = provider.GetService<IFoo>(); // The instance of Foo
    var foo3 = provider.GetService<IBar>(); // The instance of Foo

3. Check that Dispose is called when "shutting down". It will be called three times.

### Expected behavior
Since there is only one instance, I expect Dipose to be called only once.

**Comment by pakrym** [comment]
This is by design. Results of factory invocations are considered new objects and are tracked for disposal.

Duplicate of https://github.com/aspnet/Extensions/issues/694
